docs/examples/ipfs-101/main.go

```diff
@@ -0,0 +1,181 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	config "github.com/ipfs/go-ipfs-config"
+	files "github.com/ipfs/go-ipfs-files"
+	"github.com/ipfs/go-ipfs/core"
+	"github.com/ipfs/go-ipfs/core/coreapi"
+	"github.com/ipfs/go-ipfs/repo/fsrepo"
+	iCore "github.com/ipfs/interface-go-ipfs-core"
+	iCorePath "github.com/ipfs/interface-go-ipfs-core/path"
+)
+
+/// ------ Spawning the node
+
+type CfgOpt func(*config.Config)
+
+func createNode(ctx context.Context, repoPath string) (iCore.CoreAPI, error) {
+	// Open the repo
+	repo, err := fsrepo.Open(repoPath)
+	if err != nil {
+		return nil, err
+	}
+
+	// Construct the node
+	node, err := core.NewNode(ctx, &core.BuildCfg{
+		Online: true,
+		// Routing: libp2p.DHTClientOption,
+		Repo: repo,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return coreapi.NewCoreAPI(node)
+}
+
+func createTempRepo(ctx context.Context) (string, error) {
+	repoPath, err := ioutil.TempDir("", "ipfs-shell")
+	if err != nil {
+		return "", fmt.Errorf("failed to get temp dir: %s", err)
+	}
+
+	// Set default config with option for 2048 bit key
+	cfg, err := config.Init(ioutil.Discard, 2048)
+	if err != nil {
+		return "", err
+	}
+
+	// configure the temporary node
+	// cfg.Routing.Type = "dhtclient"
+	// cfg.Experimental.QUIC = true
+	// cfg.Datastore.Spec = map[string]interface{}{
+	// 	"type": "badgerds",
+	// 	"path": "badger",
+	// }
+
+	err = fsrepo.Init(repoPath, cfg)
+	if err != nil {
+		return "", fmt.Errorf("failed to init ephemeral node: %s", err)
+	}
+
+	return repoPath, nil
+}
+
+/*
+func spawnDefault(ctx context.Context) (iface.CoreAPI, error) {
+
+}
+*/
+
+/*
+func spawnEphemeral(ctx context.Context) (iface.CoreAPI, error) {
+	defaultPath, err := config.PathRoot()
+	if err != nil {
+		// shouldn't be possible
+		return nil, err
+	}
+
+	return tmpNode(ctx)
+}
+*/
+
+func spawnDefaultOrEphemeral(ctx context.Context) (iCore.CoreAPI, error) {
+	// Attempt to spawn a node in default location, check if repo already exists
+	defaultPath, err := config.PathRoot()
+	if err != nil {
+		// shouldn't be possible
+		return nil, err
+	}
+
+	ipfs, err := createNode(ctx, defaultPath)
+
+	if err == nil {
+		return ipfs, nil
+	}
+
+	// Spawn a node with a tmpRepo
+	repoPath, err := createTempRepo(ctx)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to create temp repo", err)
+	}
+
+	// Spawning an ephemeral IPFS node
+	return createNode(ctx, repoPath)
+}
+
+// ----- Writing to disk
+func writeTo(nd files.Node, fpath string) error {
```

Comment:
I took out the progress bar stuff and renamed the func I got from ipget. Which is now that I see, the same as it is on https://github.com/ipfs/go-ipfs-files/blob/96cf7c9133bdbb405d41143ced9a6161df08cc90/filewriter.go#L11. I'll double check to see i I didn't miss anything.

Reply:
I meant that, as an example, we should point users at the nice helper function so they know it exists (and don't redefine this function in all their projects).

Reply:
Updated to use the one in go-ipfs-files! Sorry for taking so much time, I honestly missed that you had pointed me at that utils package for managing files in go, I thought I was inside ipget because as you pointed out, the code is pretty much the same :D